src/ebfm/LOOP_EBM_insolation.py

```diff
@@ -141,10 +153,26 @@ def main(grid, time2, OUT):
         shade_2D = np.ones((xl, yl), dtype=int)
         shade_2D.flat[grid["ind"]] = shade
         OUT["shade"] = shade_2D.flatten()[grid["ind"]]
-    else:
-        # TODO: Shading based on lookup tables from Elmer/Ice
-        # For now: assume flat surface and no shading by surrounding terrain
-        OUT["shade"] = np.zeros_like(grid["x"])
-        OUT["shade"][elevationangle < 0] = 1
+
+    else:  # shading based on comparison of solar elevation angle with grid angle from look-up table
+        # TODO: Shading based on lookup tables from Elmer/Ice (DONE?)
+
+        # Select nearest pre-computed azimuth angle for every grid cell (1..grid.nr_az_steps)
+        az_nearest = np.rint((np.pi - azimuth) / (2.0 * np.pi) * grid["nr_az_steps"]).astype(int)
+        az_nearest[az_nearest == 0] = grid["nr_az_steps"]
+
+        # Indices into lookup table (0-based for Python)
+        ind_az = az_nearest - 1 + np.arange(grid["gpsum"], dtype=int) * grid["nr_az_steps"]
+
+        # Shading happens when the solar elevation angle <= maximum grid angle
+        temp = grid["maxgridangle"]
+        OUT["shade"] = elevationangle <= temp.flat[ind_az]
+
+        # For now: assume flat surface and no shading by surrounding terrain (no longer needed!)
+        # OUT["shade"] = np.zeros_like(grid["x"])
+        # OUT["shade"][elevationangle < 0] = 1
+
+    OUT["shade_2D"] = zeros_like(grid["x_2D"])
+    OUT["shade_2D"].flat[grid["ind"]] = OUT["shade"]
```

Comment:
Unused?

Instead of storing and duplicating these flattened `_2D` data you could also implement a small helper function similar to the following:

```
shade_2D = flatten(grid, "shade")
```

This is a general thing because I think I saw this pattern at quite a few places.